File: apps/storefront/lib/buyer-answers.ts
/**
 * The buyer's own view of the Tickets of their Sale (#315, ADR 0044; narrowed
 * by #344, ADR 0049), as rules rather than as markup.
 *
 * THE SURFACE THIS SPEAKS FOR IS ONE PAGE AND NOT TWO. The Confirmation Link
 * does not open a page of its own — it redeems into a Customer Session narrowed
 * to one Ticket Sale and lands on the Customer Area, which lists every purchase
 * the session can reach. So "the page behind the Confirmation Link" and "the
 * Customer Area" differ only in how many sales are on them, and both render
 * from this.
 *
 * WHAT THE BUYER DOES HERE, since ADR 0049: says whose each Ticket is, and
 * answers the ONE Ticket they hold. An Answer is given only by a Ticket's
 * Holder — the buyer for their Self-held Ticket (ADR 0048) — or by Event
 * Staff. Of a Ticket they do not hold the buyer sees its position, its Ticket
 * Type and its assignment state, and nothing else: no question rows, no
 * Answered/Outstanding badge, no link. The Answer Link is retired.
 *
 * TWO PAYLOADS, ONE PANEL. The sale-scoped list (`BuyerTicket`) carries every
 * Ticket of the Sale with its assignment fields and NO answer fields; the
 * held-ticket list (`HeldTicket`) carries the Tickets this Customer holds with
 * their questions and Answers and NOTHING about the Sale. The buyer's page
 * draws one row per sale-scoped Ticket and, for the row that is self-held,
 * finds its questions in the held list by `ticket_id`. The rules below are the
 * joins between the two and the two "draw nothing at all" decisions.
 *
 * IT REUSES lib/answer-link.ts RATHER THAN RESTATING IT. A Ticket Question, an
 * Answer and the body that writes one are the same shapes on every surface that
 * touches them, and the rules for turning a form into a payload are the same
 * rules — a second copy would be a second opinion about what `checked: false`
 * means, or about whether an empty field is an Answer of "".
 *
 * Pure and dependency-free — no React, no i18n runtime — so it is directly
 * unit-testable, exactly as its sibling is. It returns SHAPES and never a
 * sentence.
 *
 * WHAT IS DATA AND WHAT IS COPY. A Ticket Question's label and an Option's
 * label are read AS COINED in every Locale, like a Custom Tag (ADR 0027). Only
 * the page's own chrome follows the reader's Locale, and messages/README.md is
 * explicit that a question's words must never enter the catalog.
 */

import type { QuestionAnswer } from "@/lib/answer-link";

export type { Answer, AnswerBody, Question, QuestionAnswer, QuestionKind } from "@/lib/answer-link";

/**
 * One Ticket of the buyer's own Ticket Sale, as the sale-scoped list sends it:
 * which one it is, and whose it is.
 *
 * IT CARRIES NO QUESTION, NO ANSWER, NO OUTSTANDING COUNT AND NO LINK, for any
 * row — the Self-held one included. The API dropped them (ADR 0049), and this
 * type dropping them too is what keeps a row the buyer does not hold from ever
 * being asked to render a question it was never sent.
 */
export type BuyerTicket = {
  ticket_id: string;
  /**
   * Which of its line's units this is, 1..quantity. The buyer's handle on
   * which of four identical tickets is which until an address is given —
   * there are no seat numbers and no holder names, so "ticket 2 of 4" is the
   * whole of what can be said to tell them apart.
   */
  ordinal: number;
  ticket_type_name: string;
  /**
   * THE TICKET ASSIGNMENT (#324), and every one of these is OPTIONAL for one
   * reason: the API omits the lot while TICKET_ASSIGNMENT_ENABLED is closed, so
   * the payload a dark deployment sends is byte-identical to the one a build
   * without the feature sends. `assignment_state === undefined` is therefore how
   * this app reads the flag, and it is the only place it reads it — see
   * lib/ticket-assignment.ts, which owns every rule about these fields.
   *
   * They are declared HERE, on the row they arrive on, and interpreted THERE.
   */
  assignment_state?: string;
  /** The address this Ticket was assigned to, shown back to the buyer who typed
   * it and to nobody else on any surface. Absent while unassigned. */
  holder_email?: string;
  assigned_at?: string;
  accepted_at?: string;
  /** Whether this Ticket's Holder is the buyer themself — the one the sale
   * handed them at purchase (ADR 0048). The page says "your ticket" on it and
   * draws its questions from the held list. */
  self_held?: boolean;
  /** Whether an address may be given or changed right now, and the token saying
   * why not. */
  assignable?: boolean;
  assignable_refusal?: string;
};

/**
 * One Ticket the Customer HOLDS, as `/api/customer/held-tickets` sends it
 * (#343, ADR 0049): its questions, its Answers and what it still owes — and
 * nothing about the Sale, because the same payload serves a Holder who was
 * given the Ticket and is entitled to nothing about the purchase.
 */
export type HeldTicket = {
  ticket_id: string;
  event_name: string;
  event_slug: string;
  ticket_type_name: string;
  /** Whether Answers may still be written. False after the Event starts and on
   * a reversed Ticket Sale — and never a reason to hide what was said. */
  answerable: boolean;
  /** Why not, or "" while it is answerable. */
  answerable_refusal: string;
  /**
   * How many required Ticket Questions this Ticket still owes, from the
   * platform's ONE definition of an Outstanding Answer (#313). Never recomputed
   * on this side: a browser counting for itself would be a third opinion about
   * a debt that already, deliberately, has two.
   */
  outstanding_count: number;
  questions: QuestionAnswer[];
};

/**
 * The held-ticket row behind one of the buyer's sale-scoped rows, or null.
 *
 * NULL FOR EVERY TICKET THE BUYER DOES NOT HOLD, which is most of them, and
 * null for a Self-held Ticket whose held row has not arrived or was refused.
 * Either way the row draws as an assignment row and nothing more: a question
 * this page has not been sent is a question it has no business drawing.
 *
 * Joined on `ticket_id` and NOT on `self_held`: the flag says the buyer holds
 * it, the held list says what it asks, and a Ticket on one side and not the
 * other — the buyer gave it away between the two fetches — is drawn as an
 * ordinary assignable row rather than as a panel with no questions.
 */
export function heldRowFor(ticket: BuyerTicket, held: HeldTicket[]): HeldTicket | null {
  if (ticket.self_held !== true) return null;
  return held.find((row) => row.ticket_id === ticket.ticket_id) ?? null;
}

/**
 * Whether the buyer's one answerable row has anything to answer — i.e. whether
 * this sale has a question half at all.
 *
 * A SALE WHOSE TICKET TYPES ASK NOTHING GETS NO QUESTION HALF, not an empty
 * one. Most Organizations have never written a Ticket Question, so most
 * Customer Areas must look exactly as they did before this feature — and a
 * buyer who holds no Ticket of the Sale (gave theirs away, or bought at the
 * door) has no questions here however many the Ticket Type asks: those are
 * their Holders' to answer.
 *
 * IT SPEAKS FOR TICKET QUESTIONS ONLY, and deliberately says nothing about
 * Ticket Assignment (#324), which is behind its own flag and can be the sole
 * reason to draw the section. The caller asks both — see `saleOffersAssignment`
 * in lib/ticket-assignment.ts.
 */
export function hasAnythingToShow(tickets: BuyerTicket[], held: HeldTicket[]): boolean {
  return tickets.some((ticket) => {
    const row = heldRowFor(ticket, held);
    return row !== null && row.questions.length > 0;
  });
}

/**
 * How many Outstanding Answers the buyer owes on this sale — DEBTS AND NOT
 * TICKETS, on the one Ticket they hold. The same unit the Organization's own
 * headline figure uses, so a buyer on the phone and the member of staff they
 * are talking to are counting the same things. Zero when they hold nothing.
 */
export function saleOutstandingCount(tickets: BuyerTicket[], held: HeldTicket[]): number {
  return tickets.reduce((total, ticket) => total + (heldRowFor(ticket, held)?.outstanding_count ?? 0), 0);
}

/**
 * The held list with one row replaced by what the API just sent back from a
 * write. The held write returns ONE Ticket, not the list — each held Ticket's
 * panel stands alone — so the page patches it in by id rather than refetching.
 * A Ticket not already in the list is appended, so a write never loses a row.
 */
export function withHeldRow(held: HeldTicket[], updated: HeldTicket): HeldTicket[] {
  const index = held.findIndex((row) => row.ticket_id === updated.ticket_id);
  if (index === -1) return [...held, updated];
  return held.map((row, i) => (i === index ? updated : row));
}

File: backend/internal/catalog/service/buyer_answers.go
package service

import (
	"context"
	"time"

	"github.com/peter/ticket_pos/backend/internal/catalog"
	"github.com/peter/ticket_pos/backend/internal/catalog/repository"
)

// The buyer's own surface on the Tickets of their Sale (#315, ADR 0044; narrowed
// by #344, ADR 0049): the page behind the Confirmation Link, and the Customer
// Area for a signed-in Customer.
//
// THESE ARE ONE SURFACE AND NOT TWO. The Confirmation Link does not open a page
// of its own — it redeems into a Customer Session narrowed to one Ticket Sale
// and lands on the Customer Area, which is a single page listing every purchase
// the session can reach. So "the Confirmation Link page" and "the Customer Area"
// differ only in how many Sales the session may see, and both arrive here.
//
// WHAT THIS SURFACE IS FOR, since ADR 0049, is assignment and nothing else.
// An Answer is given only by a Ticket's Holder — the buyer for their Self-held
// Ticket, through the held-ticket routes (held_ticket_answers.go) — or by Event
// Staff. Of a Ticket the buyer does not hold they see its position, its Ticket
// Type and its assignment state, and NOT its questions, its Answers, what it
// still owes or any link that would open it. So this payload carries no
// Answer, no outstanding count and no Answer Link for ANY row, the Self-held
// one included: the buyer's own questions are read and written through the
// held-ticket routes, which is the one door every Holder uses, so there is no
// second copy of "may this person answer this Ticket" to drift.
//
// THE PREVIOUS SHAPE IS WORTH REMEMBERING, because the tests guard against it
// returning: this payload used to carry every Ticket's Answers and a per-Ticket
// Answer Link, an unauthenticated write credential over somebody else's Ticket.
// The assertion that no `answer_link`, no `questions` and no
// `outstanding_count` leave this route is in integration/buyer_answers_test.go.

// BuyerTicketAnswersView is one Ticket of the buyer's own Ticket Sale: which
// one it is, and whose it is.
//
// IT IS ITS OWN TYPE and neither TicketAnswersView nor HeldTicketAnswersView,
// for the same reason those two are separate from each other: each has a
// different reader, and a shared struct is a field added for one of them
// appearing on the other two. The name keeps the "Answers" it no longer carries
// because the route, the OpenAPI schema and the Storefront's type all spell it;
// a rename would be a diff about nothing.
type BuyerTicketAnswersView struct {
	// TicketID names which Ticket this is, so the buyer's page can post an
	// address back against it and match it to the held-ticket row when the
	// Ticket is their own. Safe: this reader proved they own the Sale.
	TicketID string `json:"ticket_id"`
	// Ordinal is which of its line's units this is, 1..quantity. It is what lets
	// the page say "ticket 2 of 4" — the only thing telling two Tickets of one
	// line apart until an address is given.
	Ordinal        int    `json:"ordinal"`
	TicketTypeName string `json:"ticket_type_name"`
	// THE TICKET ASSIGNMENT (#324, parent #322). The fields that answer the
	// buyer's question "which of my four Tickets is which".
	//
	// EVERY ONE OF THEM IS `omitempty`, AND THAT IS THE FLAG'S DOING. With
	// TICKET_ASSIGNMENT_ENABLED closed the service fills none of them, so this
	// payload is byte-identical to the one a build without the feature sends.

	// AssignmentState is `unassigned`, `assigned` or `accepted`, derived by
	// catalog.AssignmentState and never stored. Absent while the flag is closed.
	AssignmentState string `json:"assignment_state,omitempty"`
	// HolderEmail is the address this Ticket was assigned to, shown back to the
	// buyer who typed it. Empty while unassigned.
	//
	// SHOWN TO THE BUYER AND TO NOBODY ELSE ON THIS SURFACE. It is on the payload
	// because the buyer typed it and telling their four Tickets apart is the
	// whole point of the feature; it is on no public payload, where a third
	// party's address would be a disclosure.
	HolderEmail string `json:"holder_email,omitempty"`
	// AssignedAt is when this address was named, and AcceptedAt when the Holder
	// clicked. Both nil when they have not happened.
	AssignedAt *time.Time `json:"assigned_at,omitempty"`
	AcceptedAt *time.Time `json:"accepted_at,omitempty"`
	// SelfHeld is whether this Ticket's Holder is the buyer themself — the one
	// Ticket an Online Sale hands the buyer at purchase (ADR 0048), or any
	// Ticket they later assigned to their own address and accepted. The
	// Storefront says "your ticket" on it and draws its questions from the
	// held-ticket list; every other Ticket is its Holder's to answer. Absent
	// while the assignment flag is closed, like the rest of this block.
	SelfHeld bool `json:"self_held,omitempty"`
	// Assignable is whether this Ticket may be assigned or reassigned right now,
	// and AssignableRefusal names why not — a token and never a sentence,
	// because the Storefront owns the words in the reader's language.
	Assignable        bool   `json:"assignable,omitempty"`
	AssignableRefusal string `json:"assignable_refusal,omitempty"`
}

// ListBuyerTicketAnswers returns every Ticket of one of the buyer's own Ticket
// Sales with its assignment state.
//
// customerID and sessionTicketSaleID both come from the Customer Session the
// middleware validated, and NEITHER comes from the request. sessionTicketSaleID
// is empty on a full Customer Session and names one Sale on a Confirmation Link
// session; it can only ever NARROW, which is the same rule
// repository.ListTicketSalesForCustomer is held to and is what makes a forwarded
// receipt reach exactly the one purchase it was a receipt for.
//
// A Sale this session may not see resolves to an empty list rather than to a
// refusal, and that is deliberate: "you do not own this" and "this does not
// exist" must be one answer, or the endpoint becomes a way to learn which Sale
// ids are real by watching which of them refuse differently.
//
// STILL GATED ON TICKET_QUESTIONS_ENABLED, although it no longer carries a
// question: the route was born under that flag, the Storefront reads a 404 here
// as "no section", and unpicking the gate is a decision about the assignment
// feature's own flag that this narrowing does not make.
func (s *Service) ListBuyerTicketAnswers(
	ctx context.Context,
	customerID, sessionTicketSaleID, ticketSaleID string,
) ([]BuyerTicketAnswersView, error) {
	if !s.ticketQuestionsEnabled {
		return nil, catalog.ErrTicketQuestionsUnavailable()
	}
	if sessionTicketSaleID != "" && sessionTicketSaleID != ticketSaleID {
		return []BuyerTicketAnswersView{}, nil
	}

	tickets, err := s.repo.ListAnswerableTicketsForBuyer(ctx, customerID, ticketSaleID)
	if err != nil {
		return nil, err
	}
	return s.buyerTicketAnswersViews(customerID, tickets), nil
}

// buyerTicketAnswersViews assembles the buyer's rows from the scoped read.
//
// NO QUESTIONS ARE READ AND NO ANSWERS ARE LOADED, on purpose and not as an
// optimisation: a row that never carried an Answer cannot leak one. Anything the
// buyer may answer is on the held-ticket routes, behind the one "is this
// Ticket held by the caller" check.
func (s *Service) buyerTicketAnswersViews(
	customerID string,
	tickets []repository.AnswerableTicket,
) []BuyerTicketAnswersView {
	views := make([]BuyerTicketAnswersView, 0, len(tickets))
	for _, ticket := range tickets {
		view := BuyerTicketAnswersView{
			TicketID:       ticket.ID,
			Ordinal:        ticket.Ordinal,
			TicketTypeName: ticket.TicketTypeName,
		}
		s.fillBuyerAssignment(&view, customerID, ticket)
		views = append(views, view)
	}
	return views
}

// TicketSaleHasOutstandingAnswers is whether any Ticket of one Sale still owes a
// required Answer — the Sale Confirmation's one conditional sentence. False,
// not an error, while the feature is dark: a receipt is sent either way.
func (s *Service) TicketSaleHasOutstandingAnswers(ctx context.Context, ticketSaleID string) (bool, error) {
	if !s.ticketQuestionsEnabled {
		return false, nil
	}
	return s.repo.TicketSaleHasOutstandingAnswers(ctx, ticketSaleID)
}

// outstandingAnswerCount counts one Ticket's Outstanding Answers from the
// platform's ONE definition of the debt (catalog.IsOutstandingAnswer), so what a
// Holder is told they owe and what the Organization's chase list shows can
// never be two different numbers.
func outstandingAnswerCount(saleStatus string, questions []TicketQuestionAnswerView) int {
	count := 0
	for _, pair := range questions {
		if catalog.IsOutstandingAnswer(catalog.OutstandingAnswerInputs{
			Required:        pair.Question.Required,
			QuestionRetired: pair.Question.Retired,
			SaleStatus:      saleStatus,
			Answered:        pair.Answer != nil,
		}) {
			count++
		}
	}
	return count
}
